File: solutions/053.maximum-subarray/maximum-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
const maxSubArray = nums => {
  let len = nums.length

  let dp = [nums[0]]

  for (let i = 1; i < len; i++) {
    let num = nums[i]

    if (dp[i - 1] > 0) {
      dp[i] = num + dp[i - 1]
    } else {
      dp[i] = num
    }
  }

  return Math.max(...dp)
}

File: solutions/053.maximum-subarray/maximum-subarray.py
class Solution:
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if all(x < 0 for x in nums):
            return max(nums)

        ret = 0
        cur_sum= 0

        for n in nums:
            if n + cur_sum < 0:
                cur_sum = 0
            else:
                cur_sum = n + cur_sum

            ret = max(cur_sum, ret)

        return ret
